File: 0001-two-sum/0001-two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Brute Force: Time O(n^2), Space O(1)
        #for i in range(len(nums)):
        #    for j in range(i+1, len(nums)):
        #        if target == nums[i] + nums[j]:
        #            return [i, j]

        # Two-pass Hash Table: Time O(n), Space O(n)
        #hashmap = {}
        #for i in range(len(nums)):
        #    hashmap[nums[i]] = i
        #for i in range(len(nums)):
        #    complement = target - nums[i]
        #    if complement in hashmap and i != hashmap[complement]:
        #        return [i, hashmap[complement]]
                
        # One-pass Hash Table: Time O(n), Space O(n)
        #hashmap = {}
        #for i in range(len(nums)):
        #    complement = target - nums[i]
        #    if complement in hashmap:
        #        return [i, hashmap[complement]]
        #    hashmap[nums[i]] = i


        hashmap = defaultdict(int)
        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in hashmap:
                return [hashmap[complement], i]
            else:
                hashmap[nums[i]] = i

File: 0001-two-sum/0001-two-sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int,int> umap;
        for (int i = 0; i < nums.size(); i++) {
            int complement = target - nums[i];
            if (umap.find(complement) != umap.end()) {
                return {umap[complement], i};
            }
            umap[nums[i]] = i;
        }
        return {};
    }
};
//class Solution {
//public:
//    vector<int> twoSum(vector<int>& nums, int target) {
//        int n = nums.size();
//        for (int i = 0; i < n; i++) {
//            for (int j = i + 1; j < n; j++) {
//                if (nums[i] + nums[j] == target) return {i,j};
//            }
//        }
//        return {}; 
//    }
//};